bundles/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/handler/AccountHandler.java

```diff
@@ -800,4 +862,187 @@ private void handlePushActivity(@Nullable String payload) {
     void refreshAfterCommand() {
         refreshData();
     }
+
+    private @Nullable SmartHomeBaseDevice findSmartDeviceHomeJson(SmartHomeDeviceHandler handler) {
+        String id = handler.findId();
+        if (!id.isEmpty()) {
+            return jsonIdSmartHomeDeviceMapping.get(id);
+        }
+        return null;
+    }
+
+    public int shouldDiscoverSmartHomeDevices() {
+        Configuration config = getThing().getConfiguration();
+        Object discoverSmartHomeConfig = config.getProperties().get("discoverSmartHome");
+        int discoverSmartHome = 0;
+        if (discoverSmartHomeConfig instanceof Boolean) {
+            discoverSmartHome = ((boolean) discoverSmartHomeConfig) ? 2 : 0;
+        }
+        if (discoverSmartHomeConfig instanceof BigDecimal) {
+            discoverSmartHome = ((BigDecimal) discoverSmartHomeConfig).intValue();
+        }
+        return discoverSmartHome;
+    }
+
+    public boolean shouldDiscoverOpenHABSmartHomeDevices() {
+        Configuration config = getThing().getConfiguration();
+        Boolean discoverOpenHabSmartHomeDevices = (Boolean) config.getProperties()
+                .get("discoverOpenHabSmartHomeDevices");
+        if (discoverOpenHabSmartHomeDevices == null) {
+            return false;
+        }
+        return discoverOpenHabSmartHomeDevices;
+    }
+
+    public List<SmartHomeBaseDevice> updateSmartHomeDeviceList(boolean forceUpdate) {
+        Connection currentConnection = connection;
+        if (currentConnection == null) {
+            return Collections.emptyList();
+        }
+
+        if (!forceUpdate && this.smartHomeDeviceHandlers.isEmpty() && shouldDiscoverSmartHomeDevices() == 0) {
+            return Collections.emptyList();
+        }
+
+        List<SmartHomeBaseDevice> smartHomeDevices = null;
+        try {
+            if (currentConnection.getIsLoggedIn()) {
+                smartHomeDevices = currentConnection.getSmarthomeDeviceList();
+            }
+        } catch (IOException | URISyntaxException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getLocalizedMessage());
+        }
+        if (smartHomeDevices != null) {
+            // create new id map
+            Map<String, SmartHomeBaseDevice> newJsonIdSmartHomeDeviceMapping = new HashMap<>();
+            for (Object smartHomeDevice : smartHomeDevices) {
+                if (smartHomeDevice instanceof SmartHomeBaseDevice) {
+                    SmartHomeBaseDevice smartHomeBaseDevice = (SmartHomeBaseDevice) smartHomeDevice;
+                    String id = smartHomeBaseDevice.findId();
+                    if (id != null) {
+                        newJsonIdSmartHomeDeviceMapping.put(id, smartHomeBaseDevice);
+                    }
+                }
+            }
+            jsonIdSmartHomeDeviceMapping.clear();
+            jsonIdSmartHomeDeviceMapping.putAll(newJsonIdSmartHomeDeviceMapping);
```

Comment:
Why replace the contents when you can just replace the instance?
```suggestion
            jsonIdSmartHomeDeviceMapping = newJsonIdSmartHomeDeviceMapping;
```